docs/references/architecture/adr-119-dog-mempool-gossip.md

```diff
@@ -0,0 +1,287 @@
+# ADR 119: Dynamic Optimal Graph (DOG) gossip protocol
+
+## Changelog
+
+- 2024-11-04: Initial notes (@hvanz and @jmalicevic)
+- 2024-11-25: Applied comments from @hvanz and added information on decided metrics (@jmalicevic)
+
+## Status
+Accepted: [#3297][tracking-issue].
+
+
+## Context
+
+The current transaction dissemination protocol of the mempool sends each received 
+transaction to all the peers a node is connected to, except for the sender. 
+While resilient to Byzantine attacks, this type of transaction gossiping is 
+causing a lot of network traffic and nodes receiving duplicate transactions
+very frequently. 
+
+Benchmarks have also confirmed a large portion of the sent and received bytes 
+is due to transaction gossiping. 
+
+DOG is a protocol that aims to reduce the number of duplicate transactions sent
+while maintaining the resilience to attacks. 
+
+## Alternative Approaches
+
+The existing alternative approaches for transaction gossiping are:
+- FLOOD - the current transaction gossiping protocol in CometBFT
+- CAT Mempool - the transaction gossiping proposed by Celestia
+- Limiting number of peers to send a transaction to - Experimental protocol in CometBFT
+- Etherium's transaction gossiping protocol.
+
+
+### FLOOD - the current transaction gossiping protocol in CometBFT
+
+Currently, CometBFT gossips each transaction to all the connected peers, except the sender
+of the transaction. Thus, for every transaction, CometBFT keeps a list of senders to make sure
+it is not sent again to a node that has sent it. 
+This does not prevent a node to send the transaction to a node that has sent it to the 
+sender, thus creating a lot of duplicates. 
+
+This is the exact problem DOG was designed to tackle.
+
+
```

Comment:
```suggestion
### CAT protocol

Content-Addressable Transaction (CAT) is a "push-pull" gossip protocol originally implemented by Celestia on top of CometBFT's priority-mempool (aka v1, previously also in Tendermint Core), which existed in CometBFT until v0.37. 

With CAT, nodes forward ("push") transactions received from users via RPC endpoints to their peers. When a node `A` receives a transaction from another node, it will notify its peers that it has the transaction with a `SeenTx(hash(tx))` message. A node `B`, upon receiving the `SeenTx` message from `A`, will check that if the transaction is in its mempool. If not, it will "pull" the transaction from `A` by sending a `WantTx(hash(tx))` message, to which `A` responds with a `Tx` message carrying the full transaction. While `SeenTx` and `HaveTx` messages are much smaller than `Tx` messages, these additional communication steps introduce latency when disseminating transactions. The CAT protocol was especially designed for networks with low throughput and large transaction sizes. 

Efforts to port the CAT mempool to CometBFT were documented in #2027. Experimental results on a small testnet from #1472 showed that CAT effectively reduces bandwidth usage. However, its impact on latency was not evaluated in those tests. Porting CAT was finally deprioritised in favor of DOG.

```